### PR TITLE
INT4 compression

### PR DESCRIPTION
**Pull request by jane-intel**
Description: (U)INT4 data type support, (U)INT4 weights compression

Ticket: 42044

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests
* [x]  Framework operation tests **N/A**
* [x]  Transformation tests
* [x]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list **N/A**
* [x]  Guide on how to convert the **public** model **N/A**
* [x]  User guide update: **51615**

Tested for correct I4 weights compression and reading with PR https://github.com/openvinotoolkit/openvino/pull/4712 from @ilyachur 

`force_shape` attribute which is used in Model Optimizer is needed to prevent shape infer collisions after data packing as the whole feature is reducing number of stored values (two (u)int4 values are stored in one uint8 value). After we packed the value it is presented as a flattened array of np.uint8 type of reduced length, but the initial shape of data should stay in the output port of the node. `force_shape` attribute freezes this shape forever and as we do it on the late back stage it doesn't cause any troubles.

### CODE REVIEW

**Review by AlexKoff88 — COMMENTED** [review]
The code looks good but I didn't find any test that verifies its correctness.

**Comment by jane-intel** [comment]
> The code looks good but I didn't find any test that verifies its correctness.

Just added tests for (u)int4 -> uint8 data packing

**Review by AlexKoff88 — APPROVED** [review]

**Review by vgavrilo — APPROVED** [review]

File: model-optimizer/extensions/back/compress_quantized_weights.py [hunk 1]
```diff
@@ -118,7 +124,8 @@ def pattern(self):
         )
 
     @staticmethod
-    def quantize_data(fake_quantize: Node, dst_type: type):
+    def quantize_data(fake_quantize: Node, dst_type: type,
+                      quantized_type: type, mode: str):
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/back/compress_quantized_weights.py:128]: nit: merge to one line? We have a limit of 120 characters which we fit here well.

File: model-optimizer/extensions/back/compress_quantized_weights.py [hunk 2]
```diff
@@ -212,5 +224,12 @@ def replace_pattern(self, graph: Graph, match: Dict[str, Node]):
         if np.issubdtype(dst_type, np.floating):
             dst_type = data_type_str_to_np(graph.graph['cmd_params'].data_type)
 
-        self.quantize_data(fake_quantize, dst_type)
-        self.dequantize_data(fake_quantize, dst_type)
+        levels = fake_quantize.levels
+        quantized_type, mode = None, None
+        for a in sorted(self.QUANTIZATION_MAP):
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/back/compress_quantized_weights.py:229]: nit: rename variable `a` to something more meaningful

File: model-optimizer/extensions/back/compress_quantized_weights.py [hunk 3]
```diff
@@ -104,6 +104,12 @@ class CompressQuantizeWeights(BackReplacementPattern):
 
     force_clean_up = True
 
+    QUANTIZATION_MAP = {
+        # max_levels: (np_dtype, quantization_mode)
+        256: (np.int8, "signed"),
+        16: (packed_I4, "signed"),
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/back/compress_quantized_weights.py:110]: Is it possible that we need to have two ways to quantize with the same level? For example, 256 levels and both signed and unsigned mode. May be we should develop a more appropriate structure for this?

File: model-optimizer/mo/graph/port.py [hunk 4]
```diff
@@ -100,10 +100,9 @@ def _get_shape(self):
         if self.node.graph.stage == 'front':
             return None
         else:
-            if self.type == 'in':
-                return self.node.in_node(self.idx, control_flow=self.control_flow).shape
-            else:
-                return self.node.out_node(self.idx, control_flow=self.control_flow).shape
+            node_caller = self.node.in_node if self.type == 'in' else self.node.out_node
+            node = node_caller(self.idx, control_flow=self.control_flow)
+            return node.shape
```
Comment (lazarevevgeny) [inline model-optimizer/mo/graph/port.py:105]: nit: merge into one line

**Comment by lazarevevgeny** [comment]
@jane-intel , please, describe in the PR description and somewhere in the code the meaning of "force_shape" attribute.

File: model-optimizer/extensions/ops/Cast.py [hunk 5]
```diff
@@ -55,15 +111,10 @@ def infer(node: Node):
...
-                log.warning(
-                    ("{} elements of {} were clipped to zero while converting an input blob for node '{}' to {}. " +
-                     refer_to_faq_msg(77)).format(zero_match_count, new_blob.size, node.name, dst_type))
+            value = node.in_node(0).value
 
+            if dst_type in [packed_U4, packed_I4]:  # custom types conversion
+                Cast.custom_type_casting_and_packing(node, value, dst_type)
+            else:
+                node.out_port(0).data.set_value(
+                    Cast.helper_value_propagation(node.soft_get('name', node.id), value, dst_type))
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/Cast.py:120]: Introduce variable `node_name` and use it in the `assert` in the beginning of this function as well

File: model-optimizer/extensions/ops/Cast.py [hunk 6]
```diff
@@ -47,6 +49,60 @@ def type_infer(node: Node):
...
+                        " ".format(zero_match_count, new_blob.size, node_name, dst_type) + refer_to_faq_msg(77))
+        return new_blob
+
+    @staticmethod
+    def custom_type_casting_and_packing(node: Node, value, dst_type):
+        """
+        Custom types are not supported by numpy but we still need to write it to the .bin file in a compact way.
+        To do so we prepare bit representation of int4/uint4 values and store them in a numpy friendly data type.
+        We pack int4/uint4 values into uint8 type (two int4/uint4 numbers fit in uint8).
+        If the number of elements in the blob is uneven we pad them with zero value to be able to fit the bit sequence
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/Cast.py:70]: `uneven` -> `odd`

File: model-optimizer/extensions/ops/Cast.py [hunk 7]
```diff
@@ -47,6 +49,60 @@ def type_infer(node: Node):
...
+        Example: we need to represent 5 elements of int4 dtype
+            we would pad them to 6 element with the last element as zero and we would pack them into 3 uint8 values
+        """
+        assert dst_type in [packed_U4, packed_I4]
+
+        minimum_regular_dtype = np.uint8 if dst_type == packed_U4 else np.int8
+        casted_to_regular_type = Cast.helper_value_propagation(
+            node.soft_get('name', node.id), value, minimum_regular_dtype)
+
+        data_shape = node.out_port(0).data.get_shape().copy()
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/Cast.py:81]: If the result of `node.out_port(0).data.get_shape()` is `None` then we will crash here

File: model-optimizer/extensions/ops/Cast.py [hunk 8]
```diff
@@ -47,6 +49,60 @@ def type_infer(node: Node):
...
+        To do so we prepare bit representation of int4/uint4 values and store them in a numpy friendly data type.
+        We pack int4/uint4 values into uint8 type (two int4/uint4 numbers fit in uint8).
+        If the number of elements in the blob is uneven we pad them with zero value to be able to fit the bit sequence
+        into the uint8 array.
+        Example: we need to represent 5 elements of int4 dtype
+            we would pad them to 6 element with the last element as zero and we would pack them into 3 uint8 values
+        """
+        assert dst_type in [packed_U4, packed_I4]
+
+        minimum_regular_dtype = np.uint8 if dst_type == packed_U4 else np.int8
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/Cast.py:77]: Could you explain why you choose different numpy types to store `packed_U4` and `packed_I4`? Add a comment to the code.

File: model-optimizer/extensions/ops/Cast.py [hunk 9]
```diff
@@ -47,6 +49,60 @@ def type_infer(node: Node):
...
+
+        data_shape = node.out_port(0).data.get_shape().copy()
+        assert data_shape is not None
+        data_size = np.prod(data_shape)
+
+        num_values_fitting_into_uint8 = 2
+        pad = (-data_size) % num_values_fitting_into_uint8
+
+        flattened = casted_to_regular_type.flatten()
+        padded = np.array(np.concatenate((flattened, np.zeros([pad]))), dtype=minimum_regular_dtype)
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/Cast.py:89]: Should we explicitly specify data type for the `np.zeros([pad])` array? May be it is paranoid...

File: model-optimizer/extensions/ops/Cast.py [hunk 10]
```diff
@@ -47,6 +49,60 @@ def type_infer(node: Node):
...
+        assert np.prod(padded.shape) % num_values_fitting_into_uint8 == 0
+
+        num_bits = 4
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/Cast.py:92]: Variables `num_bits` and `num_values_fitting_into_uint8` are linked together. One could be calculated from another. 

nit: Also, the algorithm of padding values will work only if the `uint8`/`int8` size is evenly divided with the size of `dst_type` type. I think that we will have sizes of `dst_type` equal to a power of 2, so the current algo will work, but anyway, we can implement a better algo. Or at least, let's add a comment here about this "limitation"

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/mo/graph/port.py [hunk 11]
```diff
@@ -135,24 +135,21 @@ def _set_value(self, value):
...
-                    self.node.value = value
-                    self.node.shape = int64_array(value.shape)
+            # Set value to data node
+            data_node.value = value
+            data_node.shape = shape
+
+            # Set value to constant producer
+            if const_node.soft_get('type') == 'Const':
+                const_node.value = value
+                const_node.shape = shape
```
Comment (lazarevevgeny) [inline model-optimizer/mo/graph/port.py:152]: Please, add several unit tests for a new logic.

**Comment by lazarevevgeny** [comment]
@jane-intel , please, check the model conversion time. Make sure that we don't have performance degradation with models with huge weights.

File: model-optimizer/extensions/back/compress_quantized_weights.py (hunk 3 above)
Comment (jane-intel) [reply model-optimizer/extensions/back/compress_quantized_weights.py:110]: No, there is no such use case. It is basically a switcher if we would consider packing int4 different way than int8.

File: model-optimizer/extensions/back/compress_quantized_weights.py (hunk 1 above)
Comment (jane-intel) [reply model-optimizer/extensions/back/compress_quantized_weights.py]: Done

File: model-optimizer/extensions/back/compress_quantized_weights.py (hunk 2 above)
Comment (jane-intel) [reply model-optimizer/extensions/back/compress_quantized_weights.py]: Done

File: model-optimizer/extensions/ops/Cast.py (hunk 8 above)
Comment (jane-intel) [reply model-optimizer/extensions/ops/Cast.py:77]: Done. But we store them all in uint8. This step is initial casting from source type to the closest numpy-friendly type.

File: model-optimizer/extensions/ops/Cast.py (hunk 9 above)
Comment (jane-intel) [reply model-optimizer/extensions/ops/Cast.py]: Yes. It would be cleaner, fixed.

File: model-optimizer/extensions/ops/Cast.py (hunk 10 above)
Comment (jane-intel) [reply model-optimizer/extensions/ops/Cast.py]: Well, we have limitation stated as an assertion that dst_type is i4 or u4. Added one more assert for power of two bit num check. Lets implement better algo the day we introduce i3 / i5 data type.

File: model-optimizer/extensions/ops/Cast.py (hunk 5 above)
Comment (jane-intel) [reply model-optimizer/extensions/ops/Cast.py]: Done

File: model-optimizer/mo/graph/port.py (hunk 11 above)
Comment (jane-intel) [reply model-optimizer/mo/graph/port.py:151]: Done

**Comment by jane-intel** [comment]
> @jane-intel , please, check the model conversion time. Make sure that we don't have performance degradation with models with huge weights.

There is no significant change to test on the large number of topologies. LM1B works the same with the little deviation from one MO call to another as usual (10 calls were done per master and per branch testing).

**Review by lazarevevgeny — APPROVED** [review]

**Pull request merged**